File: Q4/src/main/java/com/yourorganization/Transport/Truck.java
package com.yourorganization.Transport; // Package for transportation classes

import java.util.List;



import java.util.ArrayList; // Importing for list functionalities

public class Truck extends Transport {
    private List<String> motorways; // Keeping a list of motorways the truck uses

    // Constructor for Truck object
    public Truck(String id, double speed, double costPerKm, List<String> motorways) {
        super(id, speed, costPerKm); // Calling the Transport's constructor
        this.motorways = new ArrayList<>(motorways); // Assigning motorways list
    }

    // Overriding the getRoutePlan method
    @Override
    public List<String> getRoutePlan() {
        return motorways;
    }

    // Getter and setter for motorways
    public List<String> getMotorways() {
        return motorways;
    }

    public void setMotorways(List<String> motorways) {
        this.motorways = motorways;
    }
}

File: Q1_Cpp/src/Transport/Truck.cpp
#include "../../include/Transport/Truck.h"
#include <sstream>
#include <algorithm>
#include <iterator>

// Constructor for Truck object
Truck::Truck(std::string id, double speed, double costPerKm, const std::vector<std::string>& motorways)
    : Transport(id, speed, costPerKm), motorways(motorways) {
}

// Overriding the getRoutePlan method from Transport
std::string Truck::getRoutePlan() {
    std::ostringstream oss;
    if (!motorways.empty()) {
        std::copy(motorways.begin(), motorways.end() - 1, std::ostream_iterator<std::string>(oss, ", "));
        oss << motorways.back();
    }
    return "Truck Route: " + oss.str(); // Creating a string of motorway routes
}

// Getter and setter for motorways
std::vector<std::string> Truck::getMotorways() const {
    return motorways;
}

void Truck::setMotorways(const std::vector<std::string>& motorways) {
    this->motorways = motorways;
}
